### PR TITLE
fix: detection of Objective-C types

### PR DESCRIPTION
**Pull request by t0rr3sp3dr0**
`ipsw` distinguished between C and Objective-C types by matching it with `[A-Z].*\*`. This is not 100% accurate, given Objective-C can start with lowercase letters and underscore.

Additionally, multi-protocol types are not being parsed correctly, causing invalid imports to be generated.

This patch addresses these problems.

1 - Before:
```objc
@import Foundation;

#include "dspd_DSPDelegate-Protocol.h"

@protocol struct vector<dspd::PortRecord<bool>, std::allocator<dspd::PortRecord<bool>>> { void * x0; void * x1; struct __compressed_pair<dspd::PortRecord<bool> *, std::allocator<dspd::PortRecord<bool>>> { void * x0; } x2; }, struct vector<dspd::PortRecord<float>, std::allocator<dspd::PortRecord<float>>> { void * x0; void * x1; struct __compressed_pair<dspd::PortRecord<float> *, std::allocator<dspd::PortRecord<float>>> { void * x0; } x2; }, {DSPDelegateDouble_Data="negotiateHandler_"{function<caulk::expected<dspd::NegotiateResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)>="__f_"{__value_func<caulk::expected<dspd::NegotiateResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)>="__buf_"{type="__lx"[32C]}"__f_"^v}}"adaptHandler_"{function<caulk::expected<dspd::AdaptResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)>="__f_"{__value_func<caulk::expected<dspd::AdaptResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)>="__buf_"{type="__lx"[32C]}"__f_"^v}}"simulateHandler_"{function<caulk::expected<dspd::SimulateResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)>="__f_"{__value_func<caulk::expected<dspd::SimulateResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)>="__buf_"{type="__lx"[32C]}"__f_"^v}}};

@interface DSPDelegateDouble : NSObject<dspd_DSPDelegate> {
  /* instance variables */
  struct DSPDelegateDouble_Data { struct function<caulk::expected<dspd::NegotiateResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)> { struct __value_func<caulk::expected<dspd::NegotiateResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)> { struct type { unsigned char x[32] __lx; } __buf_; void *__f_; } __f_; } negotiateHandler_; struct function<caulk::expected<dspd::AdaptResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)> { struct __value_func<caulk::expected<dspd::AdaptResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)> { struct type { unsigned char x[32] __lx; } __buf_; void *__f_; } __f_; } adaptHandler_; struct function<caulk::expected<dspd::SimulateResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)> { struct __value_func<caulk::expected<dspd::SimulateResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)> { struct type { unsigned char x[32] __lx; } __buf_; void *__f_; } __f_; } simulateHandler_; } admData_;
```

1 - After:
```objc
@import Foundation;

#include "dspd_DSPDelegate-Protocol.h"

@interface DSPDelegateDouble : NSObject <dspd_DSPDelegate> {
    /* instance variables */
    struct DSPDelegateDouble_Data { struct function<caulk::expected<dspd::NegotiateResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)> { struct __value_func<caulk::expected<dspd::NegotiateResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)> { struct type { unsigned char x[32] __lx; } __buf_; void *__f_; } __f_; } negotiateHandler_; struct function<caulk::expected<dspd::AdaptResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)> { struct __value_func<caulk::expected<dspd::AdaptResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)> { struct type { unsigned char x[32] __lx; } __buf_; void *__f_; } __f_; } adaptHandler_; struct function<caulk::expected<dspd::SimulateResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)> { struct __value_func<caulk::expected<dspd::SimulateResponse, std::error_code> (id<dspd_ConfigurationChangeRequest>)> { struct type { unsigned char x[32] __lx; } __buf_; void *__f_; } __f_; } simulateHandler_; } admData_;
```

2 - Before:
```objc
@import Foundation;

#include "AXPTranslationObject.h"

@protocol NSObject><NSCopying><NSSecureCoding;

@interface AXPTranslatorResponse : NSObject<NSCopying, NSSecureCoding>

@property (retain, nonatomic) NSObject<NSObject, NSCopying, NSSecureCoding> *resultData;
```

2 - After:
```objc
@import Foundation;

#include "AXPTranslationObject.h"

@interface AXPTranslatorResponse : NSObject<NSCopying, NSSecureCoding>

@property (retain, nonatomic) NSObject<NSObject, NSCopying, NSSecureCoding> *resultData;
```

3 - Before:
```objc
@import Foundation;

#include "SOExtensionContext-Protocol.h"
#include "SORemoteExtensionContextProtocol-Protocol.h"

@class ASAuthorizationProviderExtensionLoginManager;
@protocol ASAuthorizationProviderExtensionAuthorizationRequestHandler><POExtensionRegistrationProtocol;

@interface SORemoteExtensionContext : NSExtensionContext <SORemoteExtensionContextProtocol, SOExtensionContext> {
    /* instance variables */
    SOExtensionServiceConnection *_extensionServiceConnection;
    id <ASAuthorizationProviderExtensionAuthorizationRequestHandler, POExtensionRegistrationProtocol> *_extensionAuthorizationRequestHandler;
```

3 - After:
```objc
@import Foundation;

#include "SOExtensionContext-Protocol.h"
#include "SORemoteExtensionContextProtocol-Protocol.h"

@class ASAuthorizationProviderExtensionLoginManager;
@protocol ASAuthorizationProviderExtensionAuthorizationRequestHandler, POExtensionRegistrationProtocol;

@interface SORemoteExtensionContext : NSExtensionContext <SORemoteExtensionContextProtocol, SOExtensionContext> {
    /* instance variables */
    SOExtensionServiceConnection *_extensionServiceConnection;
    id <ASAuthorizationProviderExtensionAuthorizationRequestHandler, POExtensionRegistrationProtocol> *_extensionAuthorizationRequestHandler;
```

### CODE REVIEW

**Comment by blacktop** [comment]
@t0rr3sp3dr0 there are now conflicts

**Comment by t0rr3sp3dr0** [comment]
Resolved!

**Comment by t0rr3sp3dr0** [comment]
Actually, hold on. I think there is a bug in my code.

**Comment by t0rr3sp3dr0** [comment]
Done!

**Review by blacktop — COMMENTED** [review]
internal/commands/macho/objc.go:13:2: "unicode" imported and not used

**Review by blacktop — COMMENTED** [review]
make sure to run gofmt on the files etc

**Comment by blacktop** [comment]
I'm not sure this happened in this PR, but I was running and noticed that:
```objc
@class NSObject;
```

was being added now, but before I was filtering out all the forward declarations of Foundation class/protos etc due to me always @import Foundation as the top of each header?

Do you know what caused this? (I'm still debugging)

**Comment by blacktop** [comment]
I think my Foundation filtering of the forward declarations might have been left out?

**Comment by blacktop** [comment]
I'm also noticing in the forward declarations:
```objc
@protocol Proto1, id < Proto1;
```
happening?

**Comment by blacktop** [comment]
Apologies as this should have been asked when you were fixing the @optional props bug but

@optional
@property (readonly, copy) NSString *debugDescription;

Apple seems to have started marking `debugDescription` as optional recently which is why I added support (incorrectly)

however, shouldn't this prop be listed in a `@optional` section below the requird methods?

**Comment by blacktop** [comment]
> Apologies as this should have been asked when you were fixing the @optional props bug but
> 
> ```objc
> @optional
> @property (readonly, copy) NSString *debugDescription;
> ```
> 
> Apple seems to have started marking `debugDescription` as optional recently which is why I added support (incorrectly)
> 
> however, shouldn't this prop be listed in a `@optional` section below the requird methods?

I see you did fix w/ protocols but not in class headers?

**Comment by t0rr3sp3dr0** [comment]
AFAIK, `@optional` should not appear in classes, only in protocols. That’s because the class had the option to omit the optional method/property while defining its interface. If you have a variable whose type is a protocol, the underlying class may or may not have those properties/methods implemented. But if you have a class variable, either the methods/properties are implemented by that specific class or they aren’t.

But I may be wrong. Do you have an example of a class that is defining an optional property/method and doesn’t really implement it?

**Comment by t0rr3sp3dr0** [comment]
I’m going to take at the problems you mentioned later today.

**Comment by blacktop** [comment]
> AFAIK, `@optional` should not appear in classes, only in protocols. That’s because the class had the option to omit the optional method/property while defining its interface. If you have a variable whose type is a protocol, the underlying class may or may not have those properties/methods implemented. But if you have a class variable, either the methods/properties are implemented by that specific class or they aren’t.
> 
> But I may be wrong. Do you have an example of a class that is defining an optional property/method and doesn’t really implement it?

```bash
'/Users/blacktop/Downloads/ipsw_3.1.469_macOS_arm64_extras/ipsw' cd test-caches/IPSWs/21F5048f__iPhone16,2/dyld_shared_cache_arm64e Foundation --headers --output /tmp/Foundation
```

`Foundation/_NSAttributedStringFromMarkdownCreatorConcrete.h`
```objc
#ifndef _NSAttributedStringFromMarkdownCreatorConcrete_h
#define _NSAttributedStringFromMarkdownCreatorConcrete_h
@import Foundation;

@interface _NSAttributedStringFromMarkdownCreatorConcrete : NSObject<_NSAttributedStringFromMarkdownCreator> {
  /* instance variables */
  NSMutableAttributedString *_result;
}

@property (readonly) unsigned long long hash;
@property (readonly) Class superclass;
@property (readonly, copy) NSString *description;
@optional
@property (readonly, copy) NSString *debugDescription;

/* instance methods */
- (id)result;
- (void)dealloc;
- (void)appendString:(id)string with:(id)with;
- (void)appendString:(id)string;
- (id)parseExtendedAttributesFromData:(id)data error:(id *)error;
- (unsigned short)lastCharacter;
- (id)init;
- (void)finalizeWithLanguageIdentifier:(id)identifier;
@end

#endif /* _NSAttributedStringFromMarkdownCreatorConcrete_h */
```

**Comment by blacktop** [comment]
I could be mistaken, but I noticed recently (not sure if when it started) but Apple added `?` to the property attrs for `debugDescription`  I think this might be the only example of an optional Class prop?

**Comment by blacktop** [comment]
Hmm I see that `_NSAttributedStringFromMarkdownCreator` is the protocol for that Class but it's not defined in Foundation or it's deps? 🤔 that doesn't seem right?

**Comment by blacktop** [comment]
```bash
❯ 
'/Users/blacktop/Downloads/ipsw_3.1.469_macOS_arm64_extras/ipsw' dsc objc --proto test-caches/IPSWs/21F5048f__iPhone16,2/dyld_shared_cache_arm64e | grep _NSAttributedStringFromMarkdownCreator
    0x1e91da8b8: _NSAttributedStringFromMarkdownCreator Foundation
    0x1e91da8b8: _NSAttributedStringFromMarkdownCreator Foundation
    ```

**Comment by blacktop** [comment]
```bash
❯ 
'/Users/blacktop/Downloads/ipsw_3.1.469_macOS_arm64_extras/ipsw' cd test-caches/IPSWs/21F5048f__iPhone16,2/dyld_shared_cache_arm64e Foundation --proto _NSAttributedStringFromMarkdownCreator
```
```objc
@protocol _NSAttributedStringFromMarkdownCreator <NSObject>
/* instance methods */
- (void)appendString:(id)string with:(id)with;
- (void)appendString:(id)string;
- (id)parseExtendedAttributesFromData:(id)data error:(id *)error;
- (unsigned short)lastCharacter;
@optional
/* instance methods */
- (void)finalizeWithLanguageIdentifier:(id)identifier;
@end

/****************************************/
```

**Comment by blacktop** [comment]
might be filtering out files somehow?

**Comment by t0rr3sp3dr0** [comment]
About the `@class NSObject;` situation, the problem is that things like `NSObject <OS_dispatch_queue>` were not being parsed properly. In `$1 <$2>`, `$1` must be a class while `$2` must be a protocol. Previously, it was trying to match with a class and a protocol.

NSObject is kind of special because there is a class NSObject and there is also a protocol NSObject, which the class conforms to. https://opensource.apple.com/source/objc4/objc4-551.1/runtime/NSObject.h.auto.html

For some reason, `ipsw` finds 2 protocols:

```console
% ipsw cd ~/Downloads/dyld_shared_cache_x86_64h Foundation -p NSObject
@protocol NSObject

@required

@property (readonly) unsigned long long hash;
@property (readonly) Class superclass;
@property (readonly, copy) NSString *description;

/* required instance methods */
- (void)release;
- (char)isProxy;
- (id)autorelease;
- (Class)class;
- (char)conformsToProtocol:(id)protocol;
- (id)description;
- (unsigned long long)hash;
- (char)isEqual:(id)equal;
- (char)isKindOfClass:(Class)class;
- (char)isMemberOfClass:(Class)class;
- (id)performSelector:(SEL)selector;
- (id)performSelector:(SEL)selector withObject:(id)object;
- (id)performSelector:(SEL)selector withObject:(id)object withObject:(id)object;
- (char)respondsToSelector:(SEL)selector;
- (id)retain;
- (unsigned long long)retainCount;
- (id)self;
- (Class)superclass;
- (struct _NSZone *)zone;

@optional

@property (readonly, copy) NSString *debugDescription;

/* optional instance methods */
- (id)debugDescription;

@end

@protocol NSObject

@required

@property (readonly) unsigned long long hash;
@property (readonly) Class superclass;
@property (readonly, copy) NSString *description;

/* required instance methods */
- (void)release;
- (char)isProxy;
- (id)autorelease;
- (Class)class;
- (char)conformsToProtocol:(id)protocol;
- (id)description;
- (unsigned long long)hash;
- (char)isEqual:(id)equal;
- (char)isKindOfClass:(Class)class;
- (char)isMemberOfClass:(Class)class;
- (id)performSelector:(SEL)selector;
- (id)performSelector:(SEL)selector withObject:(id)object;
- (id)performSelector:(SEL)selector withObject:(id)object withObject:(id)object;
- (char)respondsToSelector:(SEL)selector;
- (id)retain;
- (unsigned long long)retainCount;
- (id)self;
- (Class)superclass;
- (struct _NSZone *)zone;

@optional

@property (readonly, copy) NSString *debugDescription;

/* optional instance methods */
- (id)debugDescription;

@end
```

And it finds something that should be the NSObject class but it definitely is not:

```console
ipsw cd ~/Downloads/dyld_shared_cache_x86_64h Foundation -c NSObject
@interface NSObjectSpecifier : NSScriptObjectSpecifier

@end
```

**Comment by t0rr3sp3dr0** [comment]
Regarding `_NSAttributedStringFromMarkdownCreator` not being found in Foundation, it looks like Apple is hiding some stuff even in public frameworks. The closest thing I could find in the framework bundled with the SDK is `__NSAttributedStringFromMarkdownWithCreator` in `Foundation.tbd`. That means that we cannot rely on `@import Foundation;` including everything that is actually present in the framework, since its TBD is missing symbols.

image

**Comment by t0rr3sp3dr0** [comment]
I plan to take a better look at these struct types in the future, as they cause compilation errors. For now, I'm keeping the old behaviour and not parsing these types. So if you have an Objective-C type that is only present inside a struct type, it won't be imported. You may also face some type redefinition errors, but that's nothing new.

**Comment by t0rr3sp3dr0** [comment]
TL;DR: we need to find out why `ipsw` isn't able to locate the symbol for the NSObject class. If that's a blocker for this PR, it's up to you.

**Comment by t0rr3sp3dr0** [comment]
Okay, it was easier than I expected. The NSObject class is defined in `libobjc`, not in `Foundation`.

```console
% ipsw cd ~/Downloads/dyld_shared_cache_x86_64h libobjc.A.dylib -c NSObject
@interface NSObject : <ROOT> <NSObject> {
    /* instance variables */
    Class isa;
}

@property (readonly) unsigned long long hash;
@property (readonly) Class superclass;
@property (readonly, copy) NSString *description;
@property (readonly, copy) NSString *debugDescription;

/* class methods */
+ (char)_isDeallocating;
+ (char)isFault;
+ (void)release;
+ (void)dealloc;
+ (char)isProxy;
+ (id)new;
+ (char)_tryRetain;
+ (id)alloc;
+ (id)allocWithZone:(struct _NSZone *)zone;
+ (char)allowsWeakReference;
+ (id)autorelease;
+ (Class)class;
+ (char)conformsToProtocol:(id)protocol;
+ (id)copy;
+ (id)copyWithZone:(struct _NSZone *)zone;
+ (id)debugDescription;
+ (id)description;
+ (void)doesNotRecognizeSelector:(SEL)selector;
+ (void)forwardInvocation:(id)invocation;
+ (id)forwardingTargetForSelector:(SEL)selector;
+ (unsigned long long)hash;
+ (id)init;
+ (void)initialize;
+ (undefined *)instanceMethodForSelector:(SEL)selector;
+ (id)instanceMethodSignatureForSelector:(SEL)selector;
+ (char)instancesRespondToSelector:(SEL)selector;
+ (char)isAncestorOfObject:(id)object;
+ (char)isEqual:(id)equal;
+ (char)isKindOfClass:(Class)class;
+ (char)isMemberOfClass:(Class)class;
+ (char)isSubclassOfClass:(Class)class;
+ (undefined *)methodForSelector:(SEL)selector;
+ (id)methodSignatureForSelector:(SEL)selector;
+ (id)mutableCopy;
+ (id)mutableCopyWithZone:(struct _NSZone *)zone;
+ (id)performSelector:(SEL)selector;
+ (id)performSelector:(SEL)selector withObject:(id)object;
+ (id)performSelector:(SEL)selector withObject:(id)object withObject:(id)object;
+ (char)resolveClassMethod:(SEL)method;
+ (char)resolveInstanceMethod:(SEL)method;
+ (char)respondsToSelector:(SEL)selector;
+ (id)retain;
+ (unsigned long long)retainCount;
+ (char)retainWeakReference;
+ (id)self;
+ (Class)superclass;
+ (struct _NSZone *)zone;

/* instance methods */
- (char)_isDeallocating;
- (char)isFault;
- (void)release;
- (void)dealloc;
- (char)isProxy;
- (char)_tryRetain;
- (char)allowsWeakReference;
- (id)autorelease;
- (Class)class;
- (char)conformsToProtocol:(id)protocol;
- (id)copy;
- (id)debugDescription;
- (id)description;
- (void)doesNotRecognizeSelector:(SEL)selector;
- (void)finalize;
- (void)forwardInvocation:(id)invocation;
- (id)forwardingTargetForSelector:(SEL)selector;
- (unsigned long long)hash;
- (id)init;
- (char)isEqual:(id)equal;
- (char)isKindOfClass:(Class)class;
- (char)isMemberOfClass:(Class)class;
- (undefined *)methodForSelector:(SEL)selector;
- (id)methodSignatureForSelector:(SEL)selector;
- (id)mutableCopy;
- (id)performSelector:(SEL)selector;
- (id)performSelector:(SEL)selector withObject:(id)object;
- (id)performSelector:(SEL)selector withObject:(id)object withObject:(id)object;
- (char)respondsToSelector:(SEL)selector;
- (id)retain;
- (unsigned long long)retainCount;
- (char)retainWeakReference;
- (id)self;
- (Class)superclass;
- (struct _NSZone *)zone;

@end
```

**Comment by blacktop** [comment]
LOL, I should have used some of the PR conversation features, as this is super hard to follow now and is entirely my fault 😩

**Comment by blacktop** [comment]
> This is what Xcode 15.3 thinks about a class that contains the `@optional` directive in its interface:
> 
> image
> 
> I think the `?` in the class property is only there to indicate that the class is implementing an optional property or method of the protocol. This is similar to the `@dynamic` directive and only describes the implementation of that class, not its interface. So we mustn't have these directives in the headers we generate.

K, I agree that `@optional` doesn't belong in Class declarations. 👍

**Comment by blacktop** [comment]
> About the `@class NSObject;` situation, the problem is that things like `NSObject <OS_dispatch_queue>` were not being parsed properly. In `$1 <$2>`, `$1` must be a class while `$2` must be a protocol. Previously, it was trying to match both `$1` and `$2` with classes + protocols.

I believe this is because there are actually 2 instances of the protocols stored in the MachO's objc proto section, before I would take care to not show duplicates, but that might have been left out when I refactored this into it's own NEW `ipsw class-dump` cmd (I'll take a look)

**Comment by blacktop** [comment]
> And it finds something that should be the NSObject class but it definitely isn't:
> 
> ```
> % ipsw cd ~/Downloads/dyld_shared_cache_x86_64h Foundation -c NSObject
> @interface NSObjectSpecifier : NSScriptObjectSpecifier
> 
> @end
> ```

This is because the "matching" is done via regex

> Regarding `_NSAttributedStringFromMarkdownCreator` not being found in Foundation, it looks like Apple is hiding some stuff even in public frameworks. The closest thing I could find in the framework bundled with the SDK is `__NSAttributedStringFromMarkdownWithCreator` in `Foundation.tbd`. That means that we cannot rely on `@import Foundation;` including everything that is actually present in the framework, since its TBD is missing symbols.
> 
> image

this might be due to this TODO #446 ?

**Comment by blacktop** [comment]
> I plan to take a better look at these struct types in the future, as they cause compilation errors. For now, I'm keeping the old behaviour and not parsing these types. So if you have an Objective-C type that is only present inside a struct type, it won't be imported. You may also face some type redefinition errors, but that's nothing new.

That would be great! It was my TODO to try and move these custom struct definitions to another .h header for Types etc

This would also clean the output up a lot and have 1 place where a user could "fix" or "properly label fields" for all uses of said structs etc

**Comment by blacktop** [comment]
> Just submitted a patch adding `libobjc.A.dylib` to the list in `scanFoundation`.
> 
> I suggest we move forward with this PR and open issues to keep track of the problems:
> 
> 1. missing symbols/headers in Foundation
> 2. implement a recursive scan of Foundation

I'll try this out and see if it works, I wonder if this is actually related to the #445 though?

**Comment by blacktop** [comment]
> Finally, I found one instance there a weird `@protocol` was being generated and I've fixed it. But I'm not sure if this also addressed the problem you were facing earlier. If you still have a problem, please let me know with which class so I can debug further.
> 
> Before:
> 
> ```objc
> @import Foundation;
> 
> @class NSObject;
> @protocol OS_dispatch_queue, OS_dispatch_queue>""__backingQueue"@"NSObject<OS_dispatch_queue>""__name"@"NSString""__nameBuffer"[300c]"__suspended"Ac"__overcommit"Ac"__propertyQoS"AC"__operationsObserverCount"AC"__operationCountObserverCount"AC"__progressReporting"AC"__mainQ"C};
> 
> @interface NSOperationQueue : NSObject <NSProgressReporting> {
>     /* instance variables */
>     struct { struct os_unfair_lock_s { unsigned int _os_unfair_lock_opaque; } __queueLock; NSOperation *__firstOperation; NSOperation *__lastOperation; NSOperation * x[6] __firstPriOperation; NSOperation * x[6] __lastPriOperation; NSMutableArray *_barriers; _NSOperationQueueProgress *_progress; atomic  __operationCount; unsigned long long x0; NSPointerArray *__activeThreads; long long __maxNumOps; int __actualMaxNumOps; int __numExecOps; NSObject<OS_dispatch_queue> *__dispatch_queue; NSObject<OS_dispatch_queue> *__backingQueue; NSString *__name; char x[300] __nameBuffer; atomic  __suspended; char x1; atomic  __overcommit; char x2; atomic  __propertyQoS; unsigned char x3; atomic  __operationsObserverCount; unsigned char x4; atomic  __operationCountObserverCount; unsigned char x5; atomic  __progressReporting; unsigned char x6; unsigned char __mainQ; } _iqp;
> ```
> 
> After:
> 
> ```objc
> @import Foundation;
> 
> @class NSObject;
> @protocol OS_dispatch_queue;
> 
> @interface NSOperationQueue : NSObject <NSProgressReporting> {
>     /* instance variables */
>     struct { struct os_unfair_lock_s { unsigned int _os_unfair_lock_opaque; } __queueLock; NSOperation *__firstOperation; NSOperation *__lastOperation; NSOperation * x[6] __firstPriOperation; NSOperation * x[6] __lastPriOperation; NSMutableArray *_barriers; _NSOperationQueueProgress *_progress; atomic  __operationCount; unsigned long long x0; NSPointerArray *__activeThreads; long long __maxNumOps; int __actualMaxNumOps; int __numExecOps; NSObject<OS_dispatch_queue> *__dispatch_queue; NSObject<OS_dispatch_queue> *__backingQueue; NSString *__name; char x[300] __nameBuffer; atomic  __suspended; char x1; atomic  __overcommit; char x2; atomic  __propertyQoS; unsigned char x3; atomic  __operationsObserverCount; unsigned char x4; atomic  __operationCountObserverCount; unsigned char x5; atomic  __progressReporting; unsigned char x6; unsigned char __mainQ; } _iqp;
> ```

This appears to be fixed (at least I can't find the example I found the other day, I should have written it down 😩 )
